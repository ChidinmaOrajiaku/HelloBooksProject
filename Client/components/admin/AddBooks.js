import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { saveImageCloudinary } from '../../actions/booksAction';
import { adminAddRequest } from '../../actions/createBooks';
import { getAllCategoryRequest } from '../../actions/getCategory';

/**
 * @class AddBooks
 */
class AddBooks extends React.Component {
  /**
     * @constructor
     * @param {object} props 
     */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      author: '',
      category: '',
      categoryData: [],
      tempImage: '',
      image: '',
      review: '',
      pointer: false,
      loader: true
    };
    this.handleChange = this.handleChange.bind(this);
    this.onAddSubmit = this.onAddSubmit.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
  }

  /**
   * 
   * @returns {nextProps} nextprops
   * @param {any} nextProps 
   * @memberof AddBooks
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.imageInputUrl && this.state.pointer) {
      this.setState({
        image: nextProps.imageInputUrl,
        pointer: false }),
      setTimeout(() => {
        this.props.adminAddRequest(this.state).then(
          () => {
            Materialize.toast(this.props.createBooksResponse, 2000, 'teal rounded');
          }
        );
      }, 1000);
    }
    this.setState({
      loader: false,
      categoryData: nextProps.getCategoryData
    });
  }

  /**
   * 
   * @returns {data} get Category
   * @memberof AddBooks
   */
  componentDidMount() {
    this.props.getAllCategoryRequest().then(() => {
      $('select').material_select();
      $('select').change(event => this.handleChange(event));
    });
  }

  /**
 * 
 * @returns {event} SyntheticEvent
 * @param {any} event 
 * @memberof AddBooks
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }
  /**
 * 
 * @constructor
 * @param {any} event 
 * @memberof AddBooks
 */
  onAddSubmit(event) {
    event.preventDefault();
    this.props.saveImageCloudinary(this.state.tempImage);
    this.setState({ pointer: true });
  }
  /**
 * 
 * @returns {event} handle Image change
 * @param {any} event 
 * @memberof AddBooks
 */
  handleImageChange(event) {
    event.preventDefault();
    const imageInput = event.target.files[0];
    const imageInputReader = new FileReader();
    if (imageInput) {
      imageInputReader.onload = () => {
        const newUpload = new Image();
        newUpload.src = imageInputReader.result;
        newUpload.onload = () => {
          this.setState({ tempImage: imageInput });
        };
      };
    }
    imageInputReader.readAsDataURL(imageInput);
  }

  /**
     * 
     * 
     * @returns {ReactElement} MarkUp
     * @memberof AddBooks
     */
  render() {
    const category = this.state.loading ? <div> Loading... </div> :
      <div className="row">
        <div className="input-field col s12 status">
          <select className="teal-text" id= "category" value="1" onChange={this.handleChange}>
            {<option className="default" value="..." disabled value> Select One </option>}
            { Object.keys(this.state.categoryData).map(key => (<option key = {key} value={this.state.categoryData[key].category }>{this.state.categoryData[key].category } </option>)
            )}
          </select>
          <label className="black-text sort">Category</label>
        </div>
      </div>;
    return (
      <div className="adminAddBooks row ontainer">
        <div className=""> <NavigationBar /> </div>
        <h4 className="col m8 offset-m3"> ADMIN UPLOAD BOOKS </h4>
        <div className="col m4 offset-m5">
          <div className="card">
            <form onSubmit={this.onAddSubmit} id="form">
              <div className="row">
                <div className="input-field col s12">
                  <input value={this.state.title} onChange={this.handleChange} id="title" type="text" className="validate" required="required"/>
                  <label htmlFor="title">Title</label>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <input value={this.state.author} onChange={this.handleChange} id="author" type="text" className="validate" required="required"/>
                  <label htmlFor="author">Author</label>
                </div>
              </div>
              {category}
              <div className="row">
                <div className="file-field input-field col s12">
                  <div className="btn file">
                    <span><i className="material-icons">file_upload</i></span>
                    <input type="file" onChange={this.handleImageChange} id="image"/>
                  </div>
                  <div className="file-path-wrapper">
                    <input className="file-path validate" type="text" placeholder="Upload Image" required="required"/>
                  </div>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <textarea value={this.state.review} onChange={this.handleChange} id="review" className="materialize-textarea" required="required"></textarea>
                  <label htmlFor="review">Review</label>
                </div>
              </div>
              <button className="btn waves-effect waves-light" type="submit" name="action">Add<i className="material-icons right">send</i></button>
            </form>
          </div>
        </div>
        <div> <Footer/></div>
      </div>
    );
  }
}

const mapStateToProps = state => (
  {
    createBooksResponse: state.createBooks[0].response.message,
    imageInputUrl: state.uploadImage[0].response || state.uploadImage[0].error,
    getCategoryData: state.getCategory[0].response
  }
);

export default connect(mapStateToProps, { adminAddRequest, saveImageCloudinary, getAllCategoryRequest })(AddBooks);
