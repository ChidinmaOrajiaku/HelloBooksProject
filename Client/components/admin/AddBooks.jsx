import React from 'react';
import ReactDOM from 'react-dom';
import { connect } from 'react-redux';
import Footer from '../Footer';
import NavigationBar from '../NavigationBar';
import { saveImageCloudinary } from '../../actions/booksAction';
import { adminAddRequest } from '../../actions/createBooks';
import { getAllCategoryRequest } from '../../actions/getCategory';

/**
 * @class AddBooks
 */
export class AddBooks extends React.Component {
  /**
     * @constructor
     * @param {object} props
     */
  constructor(props) {
    super(props);
    this.state = {
      title: '',
      author: '',
      category: '',
      categoryData: [],
      tempImage: '',
      image: '',
      imagePreview: '',
      review: '',
      pointer: false,
      loader: true
    };
    this.handleChange = this.handleChange.bind(this);
    this.onAddSubmit = this.onAddSubmit.bind(this);
    this.handleImageChange = this.handleImageChange.bind(this);
  }

  /**
   * Gets Category and mounts once component has mounted
   * @returns {object} response object
   * @memberof AddBooks
   */
  componentDidMount() {
    this.props.getAllCategoryRequest().then(() => {
      $('select').material_select();
      $('select').change(event => this.handleChange(event));
    });
    $('.modal').modal({
      dismissible: true,
      opacity: 0.5,
      inDuration: 300,
      outDuration: 200,
      startingTop: '2%',
      endingTop: '20%',
    });
  }

  /**
   *
   * @param {any} nextProps
   * @memberof AddBooks
   * @returns {nextProps} assigns nextprops to state
   */
  componentWillReceiveProps(nextProps) {
    if (nextProps.imageInputUrl && this.state.pointer) {
      this.setState({
        image: nextProps.imageInputUrl,
        pointer: false
      }),
      setTimeout(() => {
        this.props.adminAddRequest(this.state);
      }, 1000);
      setTimeout(() => {
        if (this.props.createBooksResponse.isAdded === true) {
          this.props.history.push('/books');
          Materialize.toast('Successfully Added', 2000, 'teal rounded');
        } else {
          Materialize.toast('Not Created', 2000, 'red rounded');
        }
      }, 2000);
    }
    this.setState({
      loader: false,
      categoryData: nextProps.getCategoryData
    });
  }

  /**
 *
 * @param {any} event
 * @memberof AddBooks
 * @returns {object} SyntheticEvent
 */
  handleChange(event) {
    this.setState({ [event.target.id]: event.target.value });
  }

  /**
 * Submits image to cloudinary and sets state of pointer to true
 * @param {event} event
 * @memberof AddBooks
 * @returns {object} response object
 */
  onAddSubmit(event) {
    event.preventDefault();
    this.props.saveImageCloudinary(this.state.tempImage);
    this.setState({ pointer: true });
  }

  /**
 * Handles Image change and saves image file in state
 * @param {any} event
 * @memberof AddBooks
 * @returns {object} response object
 */
  handleImageChange(event) {
    event.preventDefault();
    const imageInput = event.target.files[0];
    const imageInputReader = new FileReader();
    if (imageInput) {
      imageInputReader.onload = () => {
        const newUpload = new Image();
        newUpload.src = imageInputReader.result;
        newUpload.onload = () => {
          this.setState({
            tempImage: imageInput,
            imagePreview: newUpload.src
          });
        };
      };
    }
    imageInputReader.readAsDataURL(imageInput);
  }

  /**
     *
     * React element markup
     * @returns {object} markUp
     * @memberof AddBooks
     */
  render() {
    const preview = this.state.imagePreview === '' ?
      <button
        href="#modal1"
        className="modal-trigger previewBtn"
        disabled>Preview
      </button> :
      <button
        href="#modal1"
        className="modal-trigger previewBtn">Preview
      </button>;
    const category = this.state.loading ? <div> Loading... </div> :
      <div className="row">
        <div className="input-field col s12 status">
          <select
            className="teal-text"
            id= "category"
            value="1"
            onChange={this.handleChange}>
            {<option className="default" value="..." disabled value>
              Select One
            </option>}
            { Object.keys(this.state.categoryData).map(key =>
              (<option
                key = {key}
                value={this.state.categoryData[key].category }
              >
                {this.state.categoryData[key].category }
              </option>))}
          </select>
          <label className="black-text sort">Category</label>
        </div>
      </div>;
    return (
      <div className="adminAddBooks row ontainer">
        <div className=""> <NavigationBar /> </div>
        <h4 className="col s12 m8 offset-m3"> ADMIN UPLOAD BOOKS </h4>
        <div className="col s12 m4 offset-m5">
          <div className="card">
            <form onSubmit={this.onAddSubmit} id="form">
              <div className="row">
                <div className="input-field col s12">
                  <input
                    value={this.state.title}
                    onChange={this.handleChange}
                    id="title" type="text"
                    className="validate"
                    required="required"
                  />
                  <label htmlFor="title">Title</label>
                </div>
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <input
                    value={this.state.author}
                    onChange={this.handleChange}
                    id="author"
                    type="text"
                    className="validate"
                    required="required"
                  />
                  <label htmlFor="author">Author</label>
                </div>
              </div>
              {category}
              <div className="row">
                <div className="file-field input-field col s12">
                  <div className="btn file">
                    <span><i className="material-icons">file_upload</i></span>
                    <input
                      id="image"
                      type="file"
                      onChange={this.handleImageChange}
                    />
                  </div>
                  <div className="file-path-wrapper">
                    <input
                      className="file-path validate"
                      type="text"
                      placeholder="Upload Image"
                    />
                  </div>
                </div>
                {preview}
              </div>
              <div className="row">
                <div className="input-field col s12">
                  <textarea
                    value={this.state.review}
                    onChange={this.handleChange}
                    id="review"
                    className="materialize-textarea"
                    required="required">
                  </textarea>
                  <label htmlFor="review">Review</label>
                </div>
              </div>
              <button
                className="btn waves-effect waves-light"
                type="submit"
                name="action">Add
                <i className="material-icons right">send</i>
              </button>
            </form>
            <div id="modal1" className="modal">
              <div className="modal-content">
                <img src= {this.state.imagePreview}/>
              </div>
              <div className="modal-footer">
                <a className="modal-action modal-close">
                  <button
                    className="close">Close
                  </button></a>
              </div>
            </div>
          </div>
        </div>
        <div> <Footer /></div>
      </div>
    );
  }
}

export const mapStateToProps = state => (
  {
    createBooksResponse: state.createBooks[0],
    imageInputUrl: state.uploadImage[0].response || state.uploadImage[0].error,
    getCategoryData: state.getCategory[0].response
  }
);

export default connect(
  mapStateToProps,
  { adminAddRequest, saveImageCloudinary, getAllCategoryRequest }
)(AddBooks);
