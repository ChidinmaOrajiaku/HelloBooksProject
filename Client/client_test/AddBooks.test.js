import React from 'react';
import $ from 'jquery';
import toJson from 'enzyme-to-json';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { AddBooks } from '../components/admin/AddBooks.jsx';

configure({ adapter: new Adapter() });

describe('<AddBooks />', () => {
  const props = {
    adminAddRequest: jest.fn(() => Promise.resolve()),
    getAllCategoryRequest: jest.fn(() => Promise.resolve()),
  };

  const event = {
    preventDefault: jest.fn(),
  };
  let mountedComponent;
  const AddBooksItem = () => {
    if (!mountedComponent) {
      mountedComponent = shallow(<AddBooks {...props} />);
    }
    return mountedComponent;
  };

  beforeEach(() => {
    mountedComponent = undefined;
  });

  it('renders <AddBooks /> component', () => {
    expect(AddBooksItem()).toHaveLength(1);
  });

  it('should update state on title field change', () => {
    AddBooksItem().find('#title').simulate('change', {
      target: {
        id: 'title',
        value: 'chidebere'
      },
    });
    expect(AddBooksItem().state().title).toBe('chidebere');
  });

  it('should update state on author field change', () => {
    AddBooksItem().find('#author').simulate('change', {
      target: {
        id: 'author',
        value: 'chidebere'
      },
    });
    expect(AddBooksItem().state().author).toBe('chidebere');
  });

  it('should update state on category field change', () => {
    AddBooksItem().find('#category').simulate('change', {
      target: {
        id: 'category',
        value: 'chidebere'
      },
    });
    expect(AddBooksItem().state().category).toBe('chidebere');
  });

  it('should update state on review field change', () => {
    AddBooksItem().find('#review').simulate('change', {
      target: {
        id: 'review',
        value: 'chideberecom'
      },
    });
    expect(AddBooksItem().state().review).toBe('chideberecom');
  });

  // it('should update state on image field change', () => {
  //   AddBooksItem().find('#image').simulate('change', {
  //     target: {
  //       id: 'image',
  //       value: 'image'
  //     },
  //     event
  //   });
  //   expect(AddBooksItem().state().image).toBe('image');
  // });

  it('should match snapshot test', () => {
    const component = AddBooksItem();
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
