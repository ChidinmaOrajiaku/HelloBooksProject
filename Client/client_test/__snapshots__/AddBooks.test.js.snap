// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddBooks /> should match snapshot test 1`] = `
<div
  className="adminAddBooks row ontainer"
>
  <div
    className=""
  >
     
    <Connect(withRouter(NavigationBar)) />
     
  </div>
  <h4
    className="col m8 offset-m3"
  >
     ADMIN UPLOAD BOOKS 
  </h4>
  <div
    className="col m4 offset-m5"
  >
    <div
      className="card"
    >
      <form
        id="form"
        onSubmit={[Function]}
      >
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              id="title"
              onChange={[Function]}
              required="required"
              type="text"
              value=""
            />
            <label
              htmlFor="title"
            >
              Title
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <input
              className="validate"
              id="author"
              onChange={[Function]}
              required="required"
              type="text"
              value=""
            />
            <label
              htmlFor="author"
            >
              Author
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12 status"
          >
            <select
              className="teal-text"
              id="category"
              onChange={[Function]}
              value="1"
            >
              <option
                className="default"
                disabled={true}
                value={true}
              >
                Select One
              </option>
            </select>
            <label
              className="black-text sort"
            >
              Category
            </label>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="file-field input-field col s12"
          >
            <div
              className="btn file"
            >
              <span>
                <i
                  className="material-icons"
                >
                  file_upload
                </i>
              </span>
              <input
                id="image"
                onChange={[Function]}
                type="file"
              />
            </div>
            <div
              className="file-path-wrapper"
            >
              <input
                className="file-path validate"
                id="imageUpload"
                placeholder="Upload Image"
                required="required"
                type="text"
              />
            </div>
          </div>
          <button
            className="modal-trigger previewBtn"
            disabled={true}
            href="#modal1"
          >
            Preview
          </button>
        </div>
        <div
          className="row"
        >
          <div
            className="input-field col s12"
          >
            <textarea
              className="materialize-textarea"
              id="review"
              onChange={[Function]}
              required="required"
              value=""
            />
            <label
              htmlFor="review"
            >
              Review
            </label>
          </div>
        </div>
        <button
          className="btn waves-effect waves-light"
          name="action"
          type="submit"
        >
          Add
          <i
            className="material-icons right"
          >
            send
          </i>
        </button>
      </form>
      <div
        className="modal"
        id="modal1"
      >
        <div
          className="modal-content"
        >
          <img
            src=""
          />
        </div>
        <div
          className="modal-footer"
        >
          <a
            className="modal-action modal-close"
          >
            <button
              className="close"
            >
              Close
            </button>
          </a>
        </div>
      </div>
    </div>
  </div>
  <div>
     
    <Component />
  </div>
</div>
`;
